### ISSUE TITLE
Unable to get rollup-plugin-dts to reference local d.ts file

### ISSUE THREAD

**Issue by rdmurphy** [Swatinem/rollup-plugin-dts#71]
Hello! 👋 I'll preface this with that it may be entirely my fault this isn't working, but I've been trying to sort it out for a bit.

I have a few `npm` installed imports in a `.ts` file I've written that *do not* have types available (either locally or via DefinitelyTyped), so I've done what I normally do and have written a local `.d.ts` file to tell TypeScript to not worry about it. (This import doesn't appear in the types at all, so there's technically nothing for `rollup-plugin-dts` to "render" based on it.) My `tsconfig.json` has `strict: true` set, so it will throw an error otherwise about implicit `any`. Something like this in a file called `vendor.d.ts` that sits next to the input file:

```
declare module 'lib-without-types';
```

In VSCode and with `tsc --noEmit` this is enough — my code passes the check. But when I try to run `rollup-plugin-dts` against my `.ts` file, It throws `TS7016` about not being able to find a declaration file. As far as I can tell `rollup-plugin-dts` only considers the input script and that alone, even when the `tsconfig.json` has an `includes` value set (which should "tell" TypeScript to include my local `d.ts` file when resolving types).

(Also **super** weird — `rollup-plugin-dts` didn't care/complain about this until TypeScript 3.7 landed. If I roll back to TypeScript 3.6 it doesn't throw `TS7016`, even if I **don't** have the `vendor.d.ts` file at all!)

Is there something I'm missing or is that by design? Thank you!

**Comment by Swatinem** [comment]
Interesting case.

Have you tried using `rollup-plugin-dts` on some intermediate `.d.ts` files? It will consume those via "parsing only".

When using `.ts` files, it needs to do a full type-check (which is also slower), which might mess around a bit with these things. It tries its best to find and patch your `tsconfig.json` files on the fly, but maybe that fails somehow.

Maybe @aleclarson can chime in?

My personal preference is to just use intermediate `.d.ts` files.

**Comment by rdmurphy** [comment]
Yeah, that'd probably clear it up. That worked before I moved to using the `.ts` input.

I'm using `rollup-plugin-dts` in my [build tool library](https://github.com/rdmurphy/typebundle). The main reason I was using the `.ts` input method instead was to prevent that in-between step that'd require any extra cleanup and a separate `tsc` call — I'm using Babel to do the actual type stripping.

It gets prepared and called here:

https://github.com/rdmurphy/typebundle/blob/master/src/index.ts#L129-L175

The patching of the `tsconfig.json` was also my suspicion, and I wasn't sure it was picking up my explicit `includes` field in there. But I also have no clue how different TS gets when you're calling the methods directly like this plugin has to do. 🤔 I would have thought it would have still considered the `cwd` as a source and found my local types, but maybe not? I do know that `tsc --noEmit` doesn't get mad even when I remove the `includes` from my `tsconfig.json`.

**Comment by rdmurphy** [comment]
All right — I *think* I've recreated this outside of `rollup-plugin-dts` against the TypeScript API directly. My hunch is that this "error" was happening the entire time in my case, but for some reason 3.7 has made it become a fatal error it actually logs.

When this plugin runs `ts.createProgram` for `.ts` inputs, it only passes in the single input file it received from the Rollup options. But when `tsc` calls `ts.createProgram`, it passes a list of **all** valid files it finds in the given directory. This makes sense, because it's normally trying to compile multiple inputs. But this is **also** how it picks up any local `d.ts` files it will include as part of resolving modules it comes across.

So to make this work again, I think there would need to be some way for `rollup-plugin-dts` to find `d.ts` files it should include in `.ts` compile mode so it can also pass them to `ts.createProgram`. I could see this being as manual as a list of files someone could pass to `rollup-plugin-dts` to include, or we could figure out how to use TypeScript itself to find those `.d.ts` files to include when in `.ts` compile mode.

**Comment by Swatinem** [comment]
That makes sense. For the "normal" cases, TS is perfectly fine to lazily load its module graph from a single file you provide. I guess it fails when you have some additional `.d.ts` files.

There should be some utilities that are used to scan a directory tree, starting from a `tsconfig`. Do you want to work on a PR for that?

**Comment by rdmurphy** [comment]
I haven't had a chance yet to swing back this way but I could try and take a look. Is there already a preferred method for globbing that `rollup-plugin-dts` is using? I'd assume that both `rollup` and `typescript` itself may already have a way to do so. My gut says using whatever `typescript` does would probably make the most sense if we could find it. 🤔

**Comment by Swatinem** [comment]
Exactly, they both have different ways to glob files. And you should not mess with what rollup does, otherwise you defeat its whole purpose.

When compiling `.ts` files, the plugin will create a ts `Program` for it:

https://github.com/Swatinem/rollup-plugin-dts/blob/ac0d28aa3f04741a3c571ef26e4c58e6eda12931/src/index.ts#L81

https://github.com/Swatinem/rollup-plugin-dts/blame/master/src/program.ts#L61-L91

I can’t really remember why this is so complex, and basically has duplicate code for `createProgram`. But anyway, the `inputs` to that function is the list of files that TS considers, so your globbing should happen somewhere there.

The TS API is very badly documented sadly. I often just read the code to figure out what it does…

https://github.com/microsoft/TypeScript/blob/76ee0214f91dcbeacd43ce058f24c868f0e44def/src/executeCommandLine/executeCommandLine.ts#L495-L509

Looks like it already gets a list of files from parsing the config… Also, would be a good idea to use
an options object with named parameters instead of positional parameters in this plugin, when you touch the code.

**Comment by rdmurphy** [comment]
Okay — a breakthrough!

I discovered after poking around a bunch into `ts.parseJsonConfigFileContent` that it **does** return a `fileNames` key that lists all the files that TypeScript found in the process of reading the `tsconfig.json` file. This list *includes* any `d.ts` files it found during file resolution per what was read-in from the config, and seems to match the file resolution you see when regular `tsc` is called.

I couldn't fully wrap my head around all the cases that this plugin's `createPrograms` function is trying to account for, but if I allowed `getCompilerOptions` to return the list of `d.ts` files (by filtering out all the `.ts` files it may have found), I was able to use that list and force them into the `inputs` passed to `ts.createProgram`. Then my library could compile!

Something like this in `getCompilerOptions`:

```ts
// grabbing "fileNames" off of the parseJsonConfigFileContent return
const { fileNames, options, errors } = ts.parseJsonConfigFileContent(config, ts.sys, dirName);
  if (errors.length) {
    console.error(ts.formatDiagnostics(errors, formatHost));
    return { dirName, compilerOptions: { ...OPTIONS_OVERRIDE } };
  }
  return {
    // filter it down to only "d.ts" files using existing "dts" value
    dtsFiles: fileNames.filter(name => name.endsWith(dts)),
    dirName,
    compilerOptions: {
      ...options,
      ...OPTIONS_OVERRIDE,
    },
  };```
